onnx/defs/nn/defs.cc

```diff
@@ -1718,10 +1750,58 @@ ONNX_OPERATOR_SET_SCHEMA(
             "T",
             {"tensor(float16)", "tensor(float)", "tensor(double)"},
             "Constrain input and output types to float tensors.")
+        .TypeConstraint(
+            "T1",
+            {"tensor(bool)"},
+            "Constrain input training_mode to boolean tensors.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           propagateShapeAndTypeFromFirstInput(ctx);
-          // TODO in training mode, it may be possible to infer some of
-          // the other outputs as well.
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+
+          Dim num_channels;
+
+          unifyInputDim(ctx, 0, 1, num_channels);
+          unifyInputDim(ctx, 1, 0, num_channels);
+          unifyInputDim(ctx, 2, 0, num_channels);
+          unifyInputDim(ctx, 3, 0, num_channels);
+          unifyInputDim(ctx, 4, 0, num_channels);
+
+          if (ctx.getNumInputs() > 5 && hasInputShape(ctx, 5)) {
+            auto& mode_input_shape = getInputShape(ctx, 5);
+            // if mode is not scalar or tensor of rank 1, fail shape inference
+            if (static_cast<int>(mode_input_shape.dim_size()) != 0) {
+              if (static_cast<int>(mode_input_shape.dim_size()) > 1 ||
+                  !mode_input_shape.dim(0).has_dim_value() ||
```

Comment:
We can't produce an error-message if the input-dimension is not statically known. (But as commented above, it would be even better to just allow rank-0 tensors here.)